In [1]:
import pandas as pd
from scipy.io import arff

# Load the arff file
data = arff.loadarff('../data/raw/pc1.arff')

# Convert to a pandas DataFrame
df = pd.DataFrame(data[0])

print(df.head())

     loc   v(g)  ev(g)  iv(G)       N         V     L      D       I  \
0    1.1    1.4    1.4    1.4     1.3      1.30  1.30   1.30    1.30   
1    1.0    1.0    1.0    1.0     1.0      1.00  1.00   1.00    1.00   
2   91.0    9.0    3.0    2.0   318.0   2089.21  0.04  27.68   75.47   
3  109.0   21.0    5.0   18.0   381.0   2547.56  0.04  28.37   89.79   
4  505.0  106.0   41.0   82.0  2339.0  20696.93  0.01  75.93  272.58   

            E  ...  lOCode  lOComment  locCodeAndComment  lOBlank  uniq_Op  \
0        1.30  ...     2.0        2.0                2.0      2.0      1.2   
1        1.00  ...     1.0        1.0                1.0      1.0      1.0   
2    57833.24  ...    80.0       44.0               11.0     31.0     29.0   
3    72282.68  ...    97.0       41.0               12.0     24.0     28.0   
4  1571506.88  ...   457.0       71.0               48.0     49.0     64.0   

   uniq_Opnd  total_Op  total_Opnd  branchCount   defects  
0        1.2       1.2         1.2    

In [2]:
print("Shape:")
print(df.shape)

print("\nColumn names:")
print(df.columns.tolist())

print("\nData types:")
print(df.dtypes)

print("\nMissing values:")
print(df.isnull().sum())

Shape:
(1109, 22)

Column names:
['loc', 'v(g)', 'ev(g)', 'iv(G)', 'N', 'V', 'L', 'D', 'I', 'E', 'B', 'T', 'lOCode', 'lOComment', 'locCodeAndComment', 'lOBlank', 'uniq_Op', 'uniq_Opnd', 'total_Op', 'total_Opnd', 'branchCount', 'defects']

Data types:
loc                  float64
v(g)                 float64
ev(g)                float64
iv(G)                float64
N                    float64
V                    float64
L                    float64
D                    float64
I                    float64
E                    float64
B                    float64
T                    float64
lOCode               float64
lOComment            float64
locCodeAndComment    float64
lOBlank              float64
uniq_Op              float64
uniq_Opnd            float64
total_Op             float64
total_Opnd           float64
branchCount          float64
defects               object
dtype: object

Missing values:
loc                  0
v(g)                 0
ev(g)                0
iv(G)      

In [3]:
df["defects"].value_counts()

defects
b'false'    1032
b'true'       77
Name: count, dtype: int64

In [4]:
df["defects"].value_counts(normalize=True) * 100

defects
b'false'    93.056808
b'true'      6.943192
Name: proportion, dtype: float64

In [5]:
df.describe()

,loc,v(g),ev(g),iv(G),N,V,L,D,I,E,...,T,lOCode,lOComment,locCodeAndComment,lOBlank,uniq_Op,uniq_Opnd,total_Op,total_Opnd,branchCount
count,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1.109000e+03,...,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000,1109.000000
mean,23.376105,5.510730,2.766817,3.321371,117.393417,699.711217,0.129414,15.396267,32.904472,2.882288e+04,...,1601.273030,22.434626,4.696123,0.944995,6.745717,13.307665,20.892876,66.493417,50.901894,9.576555
std,35.284017,8.958976,5.567704,6.401998,197.336888,1509.545684,0.146952,16.337508,35.387685,1.706436e+05,...,9480.199989,33.574645,10.518447,3.345297,12.301210,8.182224,29.051434,111.703236,86.308689,16.540676
min,0.000000,1.000000,1.000000,1.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000e+00,...,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,0.000000,1.000000,0.000000,1.000000
25%,7.000000,1.000000,1.000000,1.000000,25.000000,97.670000,0.050000,5.710000,14.040000,5.581300e+02,...,31.010000,7.000000,0.000000,0.000000,1.000000,8.000000,6.000000,15.000000,10.000000,1.000000
50%,13.000000,3.000000,1.000000,2.000000,58.000000,275.100000,0.080000,11.610000,23.570000,3.189170e+03,...,177.180000,13.000000,0.000000,0.000000,2.000000,12.000000,12.000000,33.000000,24.000000,5.000000
75%,26.000000,6.000000,3.000000,3.000000,126.000000,674.040000,0.160000,20.500000,41.150000,1.238156e+04,...,687.860000,24.000000,5.000000,1.000000,8.000000,17.000000,25.000000,72.000000,56.000000,11.000000
max,602.000000,136.000000,123.000000,123.000000,2785.000000,25942.690000,2.000000,270.660000,598.330000,4.279633e+06,...,237757.400000,600.000000,159.000000,48.000000,225.000000,99.000000,538.000000,1641.000000,1144.000000,236.000000


## Initial EDA Findings
* The class distribution of the data set is imbalanced.
* The PC1 dataset consists of 1109 rows and 22 columns. Rows contain information on software modules and columns contain attributes of teh respective software modules.
* There are 22 attribute columns out of which 21 columns are used as explanatory variates (input of the model) and 1 is response variate (output of the model).
* There are no missing values found in the dataset.
* There are total of 1032 non-defective modules and 77 defective modules out of 1109 software modules.
* Several explanatory variables are rightly skewed because their means are larger than their medians.

## Day 2

In [ ]:
X = df.drop(columns=["defects"])
Y = df["defects"]

print(X.shape)
print(Y.shape)

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42, stratify=Y)

In [ ]:
print(X_train.shape)
print(X_test.shape)
print(Y_train.shape)
print(Y_test.shape)

In [ ]:
Y.value_counts(normalize=True)
Y_train.value_counts(normalize=True)
Y_test.value_counts(normalize=True)